src/satosa/micro_services/custom_routing.py

```diff
@@ -10,6 +12,66 @@
 logger = logging.getLogger(__name__)
 
 
+class DecideBackendByTarget(RequestMicroService):
+    """
+    Select which backend should be used based on who is the SAML IDP
+    """
+
+    def __init__(self, config, *args, **kwargs):
+        """
+        Constructor.
+        :param config: microservice configuration loaded from yaml file
+        :type config: Dict[str, Dict[str, str]]
+        """
+        super().__init__(*args, **kwargs)
+        self.target_mapping = config['target_mapping']
+        self.endpoint_path = config['endpoint_path']
+
+    def register_endpoints(self):
+        """
+        URL mapping of additional endpoints this micro service needs to register for callbacks.
+
+        Example of a mapping from the url path '/callback' to the callback() method of a micro service:
+            reg_endp = [
+                ("^/callback1$", self.callback),
+            ]
+
+        :rtype List[Tuple[str, Callable[[satosa.context.Context, Any], satosa.response.Response]]]
+
+        :return: A list with functions and args bound to a specific endpoint url,
+                 [(regexp, Callable[[satosa.context.Context], satosa.response.Response]), ...]
+        """
+
+        # this intercepts disco response
+        return [
+            (self.endpoint_path , self.backend_by_entityid),
+        ]
+
+    def backend_by_entityid(self, context):
+        entity_id = context.request.get('entityID')
+        tr_backend = self.target_mapping.get(entity_id)
+
+        if not all((context.request,
+                    entity_id,
+                    entity_id in self.target_mapping.keys(),
+                    tr_backend)):
+            return
```

Comment:
what is the intended behaviour?

Reply:
It's just a check, code that  I brought with me from the previous implementation.
Probably this check will return always True, so I can remove it but it's the only way to prevent exception that will block the flow (only ModuleRouting.UnknownEndpoint is handled by routing.py)

I'd like to raise ModuleRouting.UnknownEndpoint and also I'd like that UnknownEndpoint would be defined in exceptions.py. Do you agree with that?

I'll change that return with the UnknownEndpoint exception and also move this exception from ModuleRouting to exceptions.py. I can do this in this PR or in a intermediate one

Reply:
- `context.request` should always be there
- `entity_id` may be `None`
- `tr_backend` may be `None` (which means that `entity_id in self.target_mapping.keys()` also failed)

In that case you should either have a default backend (a configuration option)
or, raise an error (don't know how to handle the request).

Reply:
Ok, I'll raise the exception for these

Reply:
Exception cannot be raised ... it stops the flows.
A default backend is the answer but without a constant definition, it would rely on the preexistent if available

Reply:
that's a solution if you agree
https://github.com/IdentityPython/SATOSA/commit/d197549abc1af2fc465d23f8b3e57f4cd0bfc3f5#diff-643d6b17c79aec1dad0acafabfca8efe6dd3dcca590bcfda363aa44b06079830R54

Reply:
> Exception cannot be raised ... it stops the flows.

Yes, it stops the flow - what else should happen if you don't know how to continue with this request?

> d197549#diff-643d6b17c79aec1dad0acafabfca8efe6dd3dcca590bcfda363aa44b06079830R54

I'm not sure what to look here. This `if` has no body.

Reply:
Ok, completed here:

https://github.com/IdentityPython/SATOSA/commit/03afea4fb756b4582c27358a288ed6d051185cb3

thank you for the effort.
In the meanwhile, should we document this new feature? The triangle should be code, test and doc :)

Reply:
Roger.. we have problem ... SATOSA doesn't have any documentation on MicroServices!